src/Crdt/CrdtConfig.cs

```diff
@@ -0,0 +1,107 @@
+
+    public Action<JsonTypeInfo> MakeJsonTypeModifier()
+    {
+        return JsonTypeModifier;
+    }
+
+    public IJsonTypeInfoResolver MakeJsonTypeResolver()
+    {
+        return new DefaultJsonTypeInfoResolver
+        {
+            Modifiers = { MakeJsonTypeModifier() }
+        };
+    }
+
+    private void JsonTypeModifier(JsonTypeInfo typeInfo)
+    {
+        if (typeInfo.Type == typeof(IChange))
+        {
+            foreach (var type in ChangeTypeListBuilder.Types)
+            {
+                typeInfo.PolymorphismOptions!.DerivedTypes.Add(type);
+            }
+        }
+
+        if (typeInfo.Type == typeof(IObjectBase))
+        {
+            foreach (var type in ObjectTypeListBuilder.Types)
+            {
+                typeInfo.PolymorphismOptions!.DerivedTypes.Add(type);
+            }
+        }
+    }
+}
+
+public class ChangeTypeListBuilder
+{
+    internal List<JsonDerivedType> Types { get; } = [];
+
+    public ChangeTypeListBuilder Add<TDerived>() where TDerived : IChange, IPolyType
+    {
+        if (Types.Any(t => t.DerivedType == typeof(TDerived))) return this;
+        Types.Add(new JsonDerivedType(typeof(TDerived), TDerived.TypeName));
+        return this;
+    }
+}
+
+public class ObjectTypeListBuilder
+{
+    internal List<JsonDerivedType> Types { get; } = [];
+
+    internal List<Action<ModelBuilder, CrdtConfig>> ModelConfigurations { get; } = [];
+    public List<Action<ModelConfigurationBuilder>> ModelConventions { get; } = [];
+
+    public ObjectTypeListBuilder AddDbModelConvention(Action<ModelConfigurationBuilder> modelConvention)
+    {
+        ModelConventions.Add(modelConvention);
+        return this;
+    }
+
+    public ObjectTypeListBuilder AddDbModelConfig(Action<ModelBuilder> modelConfiguration)
+    {
+        ModelConfigurations.Add((builder, _) => modelConfiguration(builder));
+        return this;
+    }
+
+    public ObjectTypeListBuilder Add<TDerived>(Action<EntityTypeBuilder<TDerived>>? configureDb = null)
+        where TDerived : class, IObjectBase
+    {
+        if (Types.Any(t => t.DerivedType == typeof(TDerived))) throw new InvalidOperationException($"Type {typeof(TDerived)} already added");
+        Types.Add(new JsonDerivedType(typeof(TDerived), TDerived.TypeName));
+        ModelConfigurations.Add((builder, config) =>
+        {
+            if (!config.EnableProjectedTables) return;
+            var baseType = typeof(TDerived).BaseType;
+            if (baseType is not null)
+                builder.Ignore(baseType);
+            var entity = builder.Entity<TDerived>();
+            entity.HasBaseType((Type)null!);
+            entity.HasKey(e => e.Id);
+            entity.Property(e => e.Id);
+            entity.HasOne<ObjectSnapshot>()
+                .WithOne()
+                .HasForeignKey<TDerived>(ObjectSnapshot.ShadowRefName)
+            //set null otherwise it will cascade delete, which would happen whenever snapshots are deleted
```

Comment:
cascade only ever happens in one direction in SQL. take Parent -> Child, if the child is deleted the parent is never deleted, this is because the child contains the reference to the parent. However if the parent is deleted that leaves the reference in the child is invalid now which violates the FK relationship.

In this case, we delete the latest snapshot when we receive remote changes (because it's invalid as it was created before we knew about those changes). Deleting that snapshot would cause it's entity here to be deleted also, which would be fine, except that it would also cascade to other objects in the DB which reference this entity, however those may not get recreated when the snapshot that was deleted gets recreated.

Reply:
So if we don't delete those objects as part of deleting the snapshot, when **do** we delete them?

Reply:
we delete them when an object is marked as deleted see CrdtRepository.SnapshotAdded, when a snapshot is added where the entity is deleted then it will remove the projected object (projected is kinda how I'm referring to these as they're projected from the CRDT)